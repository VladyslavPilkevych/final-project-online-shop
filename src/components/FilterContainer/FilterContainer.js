import React, { memo, useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useLocation } from 'react-router-dom';
import styles from './FilterContainer.module.scss';
import Button from '../Button/Button';
import FilterPriceSlider from '../FilterPriceSlider/FilterPriceSlider';
import FilterCreatorBrand from '../FilterCreatorBrand/FilterCreatorBrand';
import swipeDown from '../../assets/icons/filterPage/swipeDown.png';
import { toggleFiltersCategories } from '../../store/actionCreators/filtersCategoriesAC';
import FilterCreatorColor from '../FilterCreatorColor/FilterCreatorColor';
import {
  filterBrand,
  newFilterProducts,
  setMinSliderValue,
  setMaxSliderValue,
  clearFilterColor,
  setFilterPaginationPage,
} from '../../store/actionCreators/filterAC';
import { repackColorsForPage } from '../../utils/repackColor';

function FilterContainer() {
  const dispatch = useDispatch();
  const location = useLocation();
  const {
    filterByColor,
    filterByBrand,
    filterCategoryProducts,
    priceSliderValues,
  } = useSelector((state) => state.filter);
  const [isOpenFilterBrands, setIsOpenFilterBrands] = useState(false);
  const [filterPrice, setFilterPrice] = useState(false);
  const [isOpenFilterColor, setIsOpenFilterColor] = useState(false);
  const [brandsFiltered, setBrandsFiltered] = useState(null);
  const [colorsFiltered, setColorsFiltered] = useState(null);

  useEffect(() => {
    if (filterCategoryProducts) {
      const brands = filterCategoryProducts.map((i) => i.name);
      setBrandsFiltered([...new Set(brands)].sort());
      setColorsFiltered(repackColorsForPage(filterCategoryProducts));
    }
  }, [filterCategoryProducts]);
  const clearFilterFn = () => {
    const priceArray = filterCategoryProducts.map(
      (item) => item.currentPrice,
    ).sort((a, b) => a - b);
    dispatch(setMinSliderValue(priceArray[0]));
    dispatch(setMaxSliderValue(priceArray[priceArray.length - 1]));
    dispatch(filterBrand([]));
    dispatch(clearFilterColor(null));
  };
  const applyFilterFn = () => {
    const filterCreators = {
      categories: location.pathname.split('/')[2],
      color: filterByColor,
      name: filterByBrand,
      minPrice: priceSliderValues.min,
      maxPrice: priceSliderValues.max,
    };
    dispatch(toggleFiltersCategories(false));
    dispatch(newFilterProducts(filterCreators));
    dispatch(setFilterPaginationPage(0));
  };
  function closeFiltersMenuOnPhone() {
    dispatch(toggleFiltersCategories(false));
  }
  function toggleToShowFilterBrands() {
    setIsOpenFilterBrands((prevState) => !prevState);
  }
  function toggleToShowFilterPrice() {
    setFilterPrice((prevState) => !prevState);
  }
  function toggleToShowFilterColor() {
    setIsOpenFilterColor((prevState) => !prevState);
  }
  return (
    <div className={styles.filterCreator}>
      <div className={styles.categoryContainers}>
        <p className={styles.filtersTopP}>Filters</p>
        <div className={styles.filtersTopPhone}>
          <p className={styles.filtersTopPhoneP}>Filter By</p>
          <div
            onClick={() => closeFiltersMenuOnPhone()}
            role="button"
            tabIndex={0}
            className={styles.filtersPhoneX}
          >
            âœ–
          </div>
        </div>
        <Button
          className={styles.clearFilterBtn}
          handleClick={clearFilterFn}
        >
          Clear Filter
        </Button>
      </div>
      <div className={styles.categoryContainers}>
        <div
          onClick={() => toggleToShowFilterBrands()}
          role="button"
          tabIndex={0}
          className={styles.twiceItems}
        >
          <p className={styles.categoryName}>Brands</p>
          <img src={swipeDown} className={styles.swiper} alt="swipeDown" />
        </div>
        {isOpenFilterBrands && brandsFiltered
          && <FilterCreatorBrand brandsFiltered={brandsFiltered} />}
      </div>
      <div className={styles.categoryContainers}>
        <div
          onClick={() => toggleToShowFilterPrice()}
          role="button"
          tabIndex={0}
          className={styles.twiceItems}
        >
          <p className={styles.categoryName}>Price</p>
          <img src={swipeDown} className={styles.swiper} alt="swipeDown" />
        </div>
        {filterPrice
          && (
            <FilterPriceSlider productsItems={filterCategoryProducts} />
          )}
      </div>
      <div className={styles.categoryContainers}>
        <div
          onClick={() => toggleToShowFilterColor()}
          role="button"
          tabIndex={0}
          className={styles.twiceItems}
        >
          <p className={styles.categoryName}>Color</p>
          <img src={swipeDown} className={styles.swiper} alt="swipeDown" />
        </div>
        {isOpenFilterColor
          && (
            <ul>
              {colorsFiltered
                && colorsFiltered.map((color) => <FilterCreatorColor key={color} color={color} />)}
            </ul>
          )}
      </div>
      <div className={styles.categoryContainers}>
        <Button
          className={[styles.applyFilterBtn, styles.applyFilterBtnActive].join(' ')}
          handleClick={() => { applyFilterFn(); }}
        >
          Apply Filters
        </Button>
      </div>
    </div>
  );
}

export default memo(FilterContainer);
