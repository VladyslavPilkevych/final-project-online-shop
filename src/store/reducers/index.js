import { combineReducers } from 'redux';
import preloaderReducer from './preloaderReducer';
import productsReducer from './productsReducer';
import menuReducer from './menuReducer';
import userReducer from './userReducer';
import searchReducer from './searchReducer';
import cartReducer from './cartReducer';
import miniMenuReducer from './miniMenuReducer';
import filtersCategoriesReducer from './filtersCategoriesReducer';

const reducer = combineReducers({
  loader: preloaderReducer,
  products: productsReducer,
  menu: menuReducer,
  miniMenu: miniMenuReducer,
  filtersCategories: filtersCategoriesReducer,
  user: userReducer,
  search: searchReducer,
  cart: cartReducer,
});

export default reducer;
