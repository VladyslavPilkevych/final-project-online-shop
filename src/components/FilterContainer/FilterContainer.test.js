import { render } from '@testing-library/react';
import FilterContainer from './FilterContainer';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import store from '../../store';

function Component() {
  return (
    <BrowserRouter>
      <Provider store={store}>
        <FilterContainer />
      </Provider>
    </BrowserRouter>
  );
}

describe('Snapshot FilterContainer should render', () => {
  test('Snapshot FilterContainer is rendered', () => {
    const { asFragment } = render(<Component />);
    expect(asFragment()).toMatchSnapshot();
  });
});
