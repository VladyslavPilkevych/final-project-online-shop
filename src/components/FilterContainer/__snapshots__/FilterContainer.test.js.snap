// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot FilterContainer should render Snapshot FilterContainer is rendered 1`] = `
<DocumentFragment>
  <div
    class="filterCreator"
  >
    <div
      class="categoryContainers"
    >
      <p
        class="filtersTopP"
      >
        Filters
      </p>
      <div
        class="filtersTopPhone"
      >
        <p
          class="filtersTopPhoneP"
        >
          Filter By
        </p>
        <div
          class="filtersPhoneX"
          role="button"
          tabindex="0"
        >
          âœ–
        </div>
      </div>
      <button
        class="clearFilterBtn [object Object]"
        type="button"
      >
        Clear Filter
      </button>
    </div>
    <div
      class="categoryContainers"
    >
      <div
        class="twiceItems"
        role="button"
        tabindex="0"
      >
        <p
          class="categoryName"
        >
          Brands
        </p>
        <img
          alt="swipeUp"
          class="swiper"
          src="swipeUp.png"
        />
      </div>
      <ul
        class="brandsContainer"
      />
    </div>
    <div
      class="categoryContainers"
    >
      <div
        class="twiceItems"
        role="button"
        tabindex="0"
      >
        <p
          class="categoryName"
        >
          Price
        </p>
        <img
          alt="swipeUp"
          class="swiper"
          src="swipeUp.png"
        />
      </div>
      <div>
        <div>
          <div
            style="display: flex; width: 100%; justify-content: space-evenly;"
          >
            <p>
              $
            </p>
            <p>
              -
            </p>
            <p>
              $
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="categoryContainers"
    >
      <div
        class="twiceItems"
        role="button"
        tabindex="0"
      >
        <p
          class="categoryName"
        >
          Color
        </p>
        <img
          alt="swipeUp"
          class="swiper"
          src="swipeUp.png"
        />
      </div>
      <ul />
    </div>
    <div
      class="categoryContainers"
    >
      <button
        class="applyFilterBtn applyFilterBtnActive [object Object]"
        type="button"
      >
        Apply Filters
      </button>
    </div>
  </div>
</DocumentFragment>
`;
